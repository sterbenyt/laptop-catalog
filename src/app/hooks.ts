import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';

// Використовуємо типізований useDispatch для безпечного диспатчу екшенів
export const useAppDispatch = () => useDispatch<AppDispatch>();

// Використовуємо типізований useSelector для отримання стану з коректною типізацією
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
