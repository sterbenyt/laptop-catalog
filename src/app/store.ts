// store.ts
import { configureStore } from '@reduxjs/toolkit';
import laptopsReducer from '../features/laptopsSlice';
import cartReducer from '../features/cartSlice'; // Додай імпорт редюсера кошика

export const store = configureStore({
    reducer: {
        laptops: laptopsReducer,
        cart: cartReducer, // Додай сюди cart
    },
});

// Тип для типізації useDispatch
export type AppDispatch = typeof store.dispatch;

// Тип для типізації useSelector
export type RootState = ReturnType<typeof store.getState>;
