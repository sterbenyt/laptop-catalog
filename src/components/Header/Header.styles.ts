import styled from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';

export const HeaderWrapper = styled.header`
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 16px 32px;
    background-color: #f8f9fa;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
`;

export const LeftSection = styled.div`
    display: flex;
    align-items: center;
    gap: 32px;
`;

export const Logo = styled.div`
    font-size: 24px;
    font-weight: bold;
`;

export const NavLinks = styled.nav`
    font-size: 18px;
    font-weight: bolder;
    display: flex;
    gap: 20px;
`;

export const NavLink = styled(RouterLink)`
    text-decoration: none;
    color: #333;
    font-weight: 500;

    &:hover {
        color: #21867a;
    }
`;

export const SearchWrapper = styled.div`
    flex: 2;
    max-width: 800px;
    margin: 0 0 0 -200px;
`;

export const CartIconWrapper = styled.div`
    position: relative;
    cursor: pointer;
    display: inline-block;
`;

export const CartCountBadge = styled.div`
  position: absolute;
  top: -10px;
  right: -10px;
  background-color: #21867a;
  color: white;
  font-size: 10px;
  font-weight: 700;
  padding: 2px 5px;
  border-radius: 10px;
  min-width: 16px;
  height: 16px;
  display: flex;
  align-items: center;
  justify-content: center;
  box-shadow: 0 0 3px rgba(0, 0, 0, 0.3);
  user-select: none;
  pointer-events: none;
`;
