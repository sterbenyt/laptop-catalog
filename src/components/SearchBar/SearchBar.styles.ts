import styled from 'styled-components';

export const SearchContainer = styled.div`
  width: 100%;
`;

export const SearchInput = styled.input`
  width: 100%;
  padding: 8px 12px;
  border-radius: 4px;
  border: 1px solid #ccc;
`;
