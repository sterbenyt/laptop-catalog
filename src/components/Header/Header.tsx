import React from 'react';
import {
    HeaderWrapper,
    NavLinks,
    Logo,
    CartIconWrapper,
    CartCountBadge,
    SearchWrapper,
    NavLink,
    LeftSection,
} from './Header.styles';
import SearchBar from '../SearchBar/SearchBar';
import cartIcon from '../../assets/cart.svg';
import { useAppSelector } from '../../app/hooks';

interface HeaderProps {
    onCartOpen?: () => void;
    onSearch?: (query: string) => void;
}

const Header: React.FC<HeaderProps> = ({ onCartOpen, onSearch }) => {
    const cartItemsCount = useAppSelector(state =>
        state.cart.items.reduce((total, item) => total + item.quantity, 0)
    );

    return (
        <HeaderWrapper>
            <LeftSection>
                <Logo>Laptop Catalog</Logo>
                <NavLinks>
                    <NavLink to="/" data-test="nav-catalog">Каталог</NavLink>
                    <NavLink to="/delivery" data-test="nav-delivery">Доставка</NavLink>
                    <NavLink to="/about" data-test="nav-about">Про нас</NavLink>
                </NavLinks>
            </LeftSection>

            <SearchWrapper>
                <SearchBar onSearch={onSearch} />
            </SearchWrapper>

            <CartIconWrapper
                onClick={onCartOpen}
                style={{ cursor: onCartOpen ? 'pointer' : 'default' }}
                data-testid="cart-icon"
            >
                <img src={cartIcon} alt="Cart" style={{ width: 24, height: 24 }} />
                {cartItemsCount > 0 && <CartCountBadge data-testid="cart-count-badge">{cartItemsCount}</CartCountBadge>}
            </CartIconWrapper>
        </HeaderWrapper>
    );
};

export default Header;
