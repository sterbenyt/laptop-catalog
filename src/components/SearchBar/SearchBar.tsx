import React, { useState, useEffect } from 'react';
import { SearchInput, SearchContainer } from './SearchBar.styles';

interface SearchBarProps {
    onSearch?: (query: string) => void;
}

const SearchBar: React.FC<SearchBarProps> = ({ onSearch }) => {
    const [query, setQuery] = useState('');

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const newQuery = e.target.value;
        setQuery(newQuery);
        if (onSearch) {
            onSearch(newQuery);
        }
    };

    return (
        <form onSubmit={(e) => e.preventDefault()} data-test="search-input">
            <SearchContainer >
                <SearchInput
                    type="text"
                    placeholder="Пошук ноутбуків..."
                    value={query}
                    onChange={handleChange}
                />
            </SearchContainer>
        </form>
    );
};

export default SearchBar;
