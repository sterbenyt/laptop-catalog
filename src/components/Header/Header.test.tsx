import React from 'react';
import { mount } from 'enzyme';
import Header from './Header';
import SearchBar from '../SearchBar/SearchBar';
import * as reduxHooks from '../../app/hooks';
import { MemoryRouter } from 'react-router-dom';

describe('Header component', () => {
    const useSelectorMock = jest.spyOn(reduxHooks, 'useAppSelector');

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly (snapshot)', () => {
        useSelectorMock.mockReturnValue(3);
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders navigation links', () => {
        useSelectorMock.mockReturnValue(0);
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        expect(wrapper.text()).toContain('Каталог');
        expect(wrapper.text()).toContain('Доставка');
        expect(wrapper.text()).toContain('Про нас');
    });

    it('renders SearchBar and passes onSearch prop', () => {
        useSelectorMock.mockReturnValue(0);
        const onSearchMock = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <Header onSearch={onSearchMock} />
            </MemoryRouter>
        );
        const searchBar = wrapper.find(SearchBar);
        expect(searchBar.exists()).toBe(true);
        expect(searchBar.prop('onSearch')).toBe(onSearchMock);
    });

    it('calls onCartOpen when cart icon is clicked', () => {
        useSelectorMock.mockReturnValue(0);
        const onCartOpenMock = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <Header onCartOpen={onCartOpenMock} />
            </MemoryRouter>
        );

        const cartIconWrapper = wrapper.find('[data-testid="cart-icon"]').first();
        expect(cartIconWrapper.exists()).toBe(true);

        cartIconWrapper.simulate('click');
        expect(onCartOpenMock).toHaveBeenCalledTimes(1);
    });

    it('shows CartCountBadge only when cartItemsCount > 0', () => {
        useSelectorMock.mockReturnValue(5);
        const wrapperWithItems = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        // Перевіряємо, що цифра 5 присутня як рядок
        expect(wrapperWithItems.text()).toContain('5');

        useSelectorMock.mockReturnValue(0);
        const wrapperWithoutItems = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        expect(wrapperWithoutItems.text()).not.toContain('5');
    });

    // Нові тести для підвищення покриття

    it('does not throw error and does not call onCartOpen if onCartOpen is not provided', () => {
        useSelectorMock.mockReturnValue(2);
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );

        const cartIconWrapper = wrapper.find('[data-testid="cart-icon"]').first();
        expect(cartIconWrapper.exists()).toBe(true);

        expect(() => {
            cartIconWrapper.simulate('click');
        }).not.toThrow();
    });



    it('sets cursor style to default if onCartOpen prop is not passed', () => {
        useSelectorMock.mockReturnValue(1);
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );

        const cartIconWrapper = wrapper.find('[data-testid="cart-icon"]').first();
        expect(cartIconWrapper.prop('style')?.cursor).toBe('default');
    });

    it('calls onSearch prop when SearchBar triggers search', () => {
        useSelectorMock.mockReturnValue({ cart: { items: [] } });
        const onSearchMock = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <Header onSearch={onSearchMock} />
            </MemoryRouter>
        );

        const searchBar = wrapper.find(SearchBar);
        expect(searchBar.exists()).toBe(true);

        const onSearchProp = searchBar.prop('onSearch');
        if (onSearchProp) {
            onSearchProp('test query');
        }
        expect(onSearchMock).toHaveBeenCalledWith('test query');
    });

    it('does not throw if onCartOpen is not provided and cart icon clicked', () => {
        useSelectorMock.mockReturnValue(2);
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        const cartIcon = wrapper.find('[data-testid="cart-icon"]').first();
        expect(() => cartIcon.simulate('click')).not.toThrow();
    });

    it('does not throw if onSearch is not provided and SearchBar triggers search', () => {
        useSelectorMock.mockReturnValue({ cart: { items: [] } });
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        const searchBar = wrapper.find(SearchBar);
        expect(() => {
            const onSearchProp = searchBar.prop('onSearch');
            if (onSearchProp) onSearchProp('query');
        }).not.toThrow();
    });

    it('renders correctly without any props', () => {
        useSelectorMock.mockReturnValue(0);
        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        expect(wrapper.exists()).toBe(true);
    });

    it('correctly calculates cartItemsCount from items array', () => {
        useSelectorMock.mockImplementation(selector =>
            selector({
                laptops: {} as any,  // Якщо хочеш швидко заглушити
                cart: {
                    items: [
                        { id: '1', quantity: 2, title: 'Item 1', price: 100 },
                        { id: '2', quantity: 3, title: 'Item 2', price: 200 },
                    ],
                    isOpen: false,
                },
            })
        );

        const wrapper = mount(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );

        const badges = wrapper.find('[data-testid="cart-count-badge"]');
        expect(badges.length).toBeGreaterThan(0); // перевірка що є
        const badge = badges.at(0); // беремо перший елемент
        expect(badge.text()).toBe('5'); // 2 + 3 = 5
    });


});
