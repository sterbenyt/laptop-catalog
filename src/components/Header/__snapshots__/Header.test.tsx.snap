// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correctly (snapshot) 1`] = `ReactWrapper {}`;
